pkg/controller/daemon/util/daemonset_util.go

```diff
@@ -133,3 +134,60 @@ func SplitByAvailablePods(minReadySeconds int32, pods []*v1.Pod) ([]*v1.Pod, []*
 	}
 	return availablePods, unavailablePods
 }
+
+// AppendDaemonSetPodNodeAffinity adds a NodeAffinity term with the given "nodeName" to the "affinity" terms.
+func AppendDaemonSetPodNodeAffinity(affinity *v1.Affinity, nodename string) *v1.Affinity {
+	nodeSelector := &v1.NodeSelector{
+		NodeSelectorTerms: []v1.NodeSelectorTerm{
+			{
+				MatchExpressions: []v1.NodeSelectorRequirement{
+					{
+						Key:      kubeletapis.LabelHostname,
+						Operator: v1.NodeSelectorOpIn,
+						Values:   []string{nodename},
+					},
+				},
+			},
+		},
+	}
+
+	if affinity == nil {
+		return &v1.Affinity{
+			NodeAffinity: &v1.NodeAffinity{
+				RequiredDuringSchedulingIgnoredDuringExecution: nodeSelector,
+			},
+		}
+	}
+
+	if affinity.NodeAffinity == nil {
+		affinity.NodeAffinity = &v1.NodeAffinity{
+			RequiredDuringSchedulingIgnoredDuringExecution: nodeSelector,
+		}
+		return affinity
+	}
+
+	nodeAffinity := affinity.NodeAffinity
+
+	if nodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil {
+		nodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution = nodeSelector
+		return affinity
+	}
+
+	// Removes hostname node selector, as only the target hostname will take effect.
+	for _, term := range nodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms {
```

Comment:
This tells the DaemonSet controller to ignore all hostname related RequiredDuringSchedulingIgnoredDuringExecution node affinity specified in its template. 

DaemonSet controller shouldn't do that. When it schedules a pod, it takes node affinity into account, instead of forcing all its pod to be scheduled on each node (by setting `nodeName`).

Reply:
Synced offline, the input `nodename` shouldn't have conflict with `affinity` because `nodeSelectionPredicates` is invoked first